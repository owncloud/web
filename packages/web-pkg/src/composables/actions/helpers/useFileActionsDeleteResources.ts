import { cloneStateObject } from '../../../helpers/store'
import { isSameResource } from '../../../helpers/resource'
import { Resource, SpaceResource } from '@ownclouders/web-client'
import { isLocationSpacesActive } from '../../../router'
import { dirname } from 'path'
import { createFileRouteOptions } from '../../../helpers'
import { computed, unref } from 'vue'
import { queryItemAsString } from '../../appDefaults'
import { useGetMatchingSpace } from '../../spaces'
import { useRouteQuery } from '../../router'
import { useClientService } from '../../clientService'
import { useRouter } from '../../router'
import { useGettext } from 'vue3-gettext'
import { ref } from 'vue'
import {
  useMessages,
  useModals,
  useSpacesStore,
  useConfigStore,
  useResourcesStore,
  useSharesStore
} from '../../piniaStores'
import { storeToRefs } from 'pinia'
import { useDeleteWorker } from '../../webWorkers'

export const useFileActionsDeleteResources = () => {
  const configStore = useConfigStore()
  const messageStore = useMessages()
  const router = useRouter()
  const language = useGettext()
  const { getMatchingSpace } = useGetMatchingSpace()
  const { $gettext, $ngettext } = language
  const clientService = useClientService()
  const { dispatchModal } = useModals()
  const spacesStore = useSpacesStore()
  const sharesStore = useSharesStore()
  const { startWorker } = useDeleteWorker({
    concurrentRequests: configStore.options.concurrentRequests.resourceBatchActions
  })

  const resourcesStore = useResourcesStore()
  const { currentFolder } = storeToRefs(resourcesStore)

  const resourcesToDelete = ref<Resource[]>([])

  const currentPageQuery = useRouteQuery('page', '1')
  const currentPage = computed(() => {
    return parseInt(queryItemAsString(unref(currentPageQuery)))
  })

  const itemsPerPageQuery = useRouteQuery('items-per-page', '1')
  const itemsPerPage = computed(() => {
    return parseInt(queryItemAsString(unref(itemsPerPageQuery)))
  })

  const resources = computed(() => {
    return cloneStateObject<Resource[]>(unref(resourcesToDelete))
  })

  const dialogTitle = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'
    let title: string = null

    if (currentResources.length === 1) {
      if (isFolder) {
        title = $gettext(
          'Permanently delete folder "%{name}"',
          {
            name: currentResources[0].name
          },
          true
        )
      } else {
        title = $gettext(
          'Permanently delete file "%{name}"',
          {
            name: currentResources[0].name
          },
          true
        )
      }
      return title
    }

    return $ngettext(
      'Permanently delete selected resource?',
      'Permanently delete %{amount} selected resources?',
      currentResources.length,
      { amount: currentResources.length.toString() },
      false
    )
  })

  const dialogMessage = computed(() => {
    const currentResources = unref(resources)
    const isFolder = currentResources[0].type === 'folder'

    if (currentResources.length === 1) {
      if (isFolder) {
        return $gettext(
          'Are you sure you want to delete this folder? All its content will be permanently removed. This action cannot be undone.'
        )
      }
      return $gettext(
        'Are you sure you want to delete this file? All its content will be permanently removed. This action cannot be undone.'
      )
    }

    return $gettext(
      'Are you sure you want to delete all selected resources? All their content will be permanently removed. This action cannot be undone.'
    )
  })

  const trashbin_delete = (space: SpaceResource) => {
    const originalRoute = unref(router.currentRoute)

    startWorker(
      { topic: 'trashBinDelete', space, resources: unref(resources) },
      ({ successful, failed }) => {
        if (successful.length) {
          const title =
            successful.length === 1 && unref(resources).length === 1
              ? $gettext('"%{item}" was deleted successfully', { item: successful[0].name })
              : $ngettext(
                  '%{itemCount} item was deleted successfully',
                  '%{itemCount} items were deleted successfully',
                  successful.length,
                  { itemCount: successful.length.toString() },
                  true
                )

          messageStore.showMessage({ title })
        }

        failed.forEach(({ resource }) => {
          const title = $gettext('Failed to delete "%{item}"', { item: resource.name })
          messageStore.showErrorMessage({ title, errors: [new Error()] })
        })

        // user hasn't navigated to another location meanwhile
        if (
          originalRoute.name === unref(router.currentRoute).name &&
          originalRoute.query?.fileId === unref(router.currentRoute).query?.fileId
        ) {
          resourcesStore.removeResources(successful)
          resourcesStore.resetSelection()
        }
      }
    )
  }

  const filesList_delete = (resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    const resourceSpaceMapping = unref(resources).reduce<
      Record<string, { space: SpaceResource; resources: Resource[] }>
    >((acc, resource) => {
      if (resource.storageId in acc) {
        acc[resource.storageId].resources.push(resource)
        return acc
      }

      const matchingSpace = getMatchingSpace(resource)

      if (!(matchingSpace.id in acc)) {
        acc[matchingSpace.id] = { space: matchingSpace, resources: [] }
      }

      acc[matchingSpace.id].resources.push(resource)
      return acc
    }, {})

    const originalCurrentFolderId = unref(currentFolder)?.id

    return Object.values(resourceSpaceMapping).map(
      ({ space: spaceForDeletion, resources: resourcesForDeletion }) => {
        startWorker(
          { topic: 'fileListDelete', space: spaceForDeletion, resources: resourcesForDeletion },
          async ({ successful, failed }) => {
            if (successful.length) {
              const title =
                successful.length === 1 && resources.length === 1
                  ? $gettext('"%{item}" was moved to trash bin', { item: successful[0].name })
                  : $ngettext(
                      '%{itemCount} item was moved to trash bin',
                      '%{itemCount} items were moved to trash bin',
                      successful.length,
                      { itemCount: successful.length.toString() },
                      true
                    )

              messageStore.showMessage({ title })
            }

            failed.forEach(({ error, resource }) => {
              let title = $gettext('Failed to delete "%{resource}"', { resource: resource.name })
              if (error.statusCode === 423) {
                title = $gettext('Failed to delete "%{resource}" - the file is locked', {
                  resource: resource.name
                })
              }

              messageStore.showErrorMessage({ title, errors: [error] })
            })

            // user hasn't navigated to another location meanwhile
            if (originalCurrentFolderId === unref(currentFolder)?.id) {
              resourcesStore.removeResources(successful)

              successful.forEach(({ id }) => {
                resourcesStore.removeSelection(id)
              })

              const activeFilesCount = resourcesStore.activeResources.length
              const pageCount = Math.ceil(unref(activeFilesCount) / unref(itemsPerPage))
              if (unref(currentPage) > 1 && unref(currentPage) > pageCount) {
                // reset pagination to avoid empty lists (happens when deleting all items on the last page)
                currentPageQuery.value = pageCount.toString()
              }
            }

            // Load quota
            if (
              isLocationSpacesActive(router, 'files-spaces-generic') &&
              !['public', 'share'].includes(spaceForDeletion?.driveType)
            ) {
              const graphClient = clientService.graphAuthenticated
              const updatedSpace = await graphClient.drives.getDrive(
                unref(resources)[0].storageId,
                sharesStore.graphRoles
              )
              spacesStore.updateSpaceField({
                id: updatedSpace.id,
                field: 'spaceQuota',
                value: updatedSpace.spaceQuota
              })
            }

            if (
              unref(resourcesToDelete).length &&
              isSameResource(unref(resourcesToDelete)[0], unref(currentFolder))
            ) {
              // current folder is being deleted
              return router.push(
                createFileRouteOptions(spaceForDeletion, {
                  path: dirname(unref(resourcesToDelete)[0].path),
                  fileId: unref(resourcesToDelete)[0].parentFolderId
                })
              )
            }
          }
        )
      }
    )
  }

  const displayDialog = (space: SpaceResource, resources: Resource[]) => {
    resourcesToDelete.value = [...resources]

    dispatchModal({
      variation: 'danger',
      title: unref(dialogTitle),
      message: unref(dialogMessage),
      confirmText: $gettext('Delete'),
      onConfirm: () => trashbin_delete(space)
    })
  }

  return {
    displayDialog,
    filesList_delete
  }
}
