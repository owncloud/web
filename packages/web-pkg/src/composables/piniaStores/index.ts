export * from './apps'
export * from './auth'
export * from './capabilities'
export * from './clipboard'
export * from './config'
export * from './extensionRegistry'
export * from './messages'
export * from './modals'
export * from './resources'
export * from './shares'
export * from './spaces'
export * from './theme'
export * from './user'
export * from './webWorkers'
export * from './inviteTokensList'
export * from './app'
