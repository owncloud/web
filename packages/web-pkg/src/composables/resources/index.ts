export * from './useCanBeOpenedWithSecureView'
export * from './useGetResourceContext'
export * from './useLoadFileInfoById'
export * from './useResourceContents'
export * from './useResourcePath'
