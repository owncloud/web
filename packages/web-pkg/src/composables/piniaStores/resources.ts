import { defineStore } from 'pinia'
import { Ref, computed, ref, unref } from 'vue'
import {
  extractStorageId,
  isMountPointSpaceResource,
  isProjectSpaceResource,
  SpaceResource,
  type Resource
} from '@ownclouders/web-client'
import { getIndicators, getParentPaths } from '../../helpers'
import { AncestorMetaData, AncestorMetaDataValue } from '../../types'
import { DavProperty, WebDAV } from '@ownclouders/web-client/webdav'
import { useSpacesStore } from './spaces'
import { useUserStore } from './user'
import { useConfigStore } from './config'

export const useResourcesStore = defineStore('resources', () => {
  const configStore = useConfigStore()
  const spacesStore = useSpacesStore()
  const userStore = useUserStore()

  const resources = ref<Resource[]>([]) as Ref<Resource[]>
  const currentFolder = ref<Resource>()
  const ancestorMetaData = ref<AncestorMetaData>({})

  const activeResources = computed(() => {
    let res = unref(resources)

    if (!unref(areHiddenFilesShown)) {
      res = res.filter((file) => !file.name.startsWith('.'))
    }

    return res
  })

  const totalResourcesCount = computed(() => {
    const fileCount = unref(resources).filter(({ type }) => type === 'file').length
    const hiddenFileCount = unref(resources).filter(
      ({ type, name }) => type === 'file' && name.startsWith('.')
    ).length
    const folderCount = unref(resources).filter(({ type }) => type === 'folder').length
    const hiddenFolderCount = unref(resources).filter(
      ({ type, name }) => type === 'folder' && name.startsWith('.')
    ).length
    const spaceCount = unref(resources).filter(isProjectSpaceResource).length

    return {
      files: fileCount,
      hiddenFiles: hiddenFileCount,
      folders: folderCount,
      hiddenFolders: hiddenFolderCount,
      spaces: spaceCount
    }
  })

  const setResources = (file: Resource[]) => {
    resources.value = file
  }

  const removeResources = (values: Resource[]) => {
    resources.value = unref(resources).filter((file) => !values.find(({ id }) => id === file.id))
  }

  const clearResources = () => {
    resources.value = []
  }

  const upsertResource = (resource: Resource) => {
    const existing = unref(resources).find(({ id }) => id === resource.id)
    if (existing) {
      Object.assign(existing, resource)
      return
    }
    unref(resources).push(resource)
  }

  const upsertResources = (values: Resource[]) => {
    const other = unref(resources).filter((f) => !values.some((r) => r.path === f.path))
    resources.value = [...other, ...values]
  }

  const updateResourceField = <T extends Resource>({
    id,
    field,
    value
  }: {
    id: T['id']
    field: keyof T
    value: T[keyof T]
  }) => {
    const resource = unref(resources).find((resource) => id === resource.id) as T
    if (resource) {
      resource[field] = value
    }
  }

  const setCurrentFolder = (value: Resource) => {
    currentFolder.value = value
  }

  const clearCurrentFolder = () => {
    currentFolder.value = undefined
  }

  const initResourceList = <T extends Resource>(data: { resources: T[]; currentFolder: T }) => {
    resources.value = data.resources
    currentFolder.value = data.currentFolder
  }

  const clearResourceList = () => {
    resources.value = []
    currentFolder.value = undefined
    selectedIds.value = []
  }

  const selectedIds = ref<string[]>([])
  const latestSelectedId = ref<string>(null)

  const selectedResources = computed(() => {
    return unref(resources).filter((f) => unref(selectedIds).includes(f.id))
  })

  const setSelection = (ids: string[]) => {
    const latestSelected = ids.find((id) => !unref(selectedIds).includes(id))

    if (latestSelected) {
      latestSelectedId.value = latestSelected
    }
    selectedIds.value = ids
  }

  const addSelection = (id: string) => {
    latestSelectedId.value = id
    if (!unref(selectedIds).includes(id)) {
      unref(selectedIds).push(id)
    }
  }

  const removeSelection = (id: string) => {
    latestSelectedId.value = id
    if (unref(selectedIds).includes(id)) {
      selectedIds.value = unref(selectedIds).filter((i) => i !== id)
    }
  }

  const toggleSelection = (id: string) => {
    if (unref(selectedIds).includes(id)) {
      removeSelection(id)
    } else {
      addSelection(id)
    }
  }

  const resetSelection = () => {
    selectedIds.value = []
  }

  const setLastSelectedId = (id: string) => {
    latestSelectedId.value = id
  }

  const areHiddenFilesShown = ref(true)
  const areFileExtensionsShown = ref(true)
  const areWebDavDetailsShown = ref(false)

  const setAreHiddenFilesShown = (value: boolean) => {
    areHiddenFilesShown.value = value
    window.localStorage.setItem('oc_hiddenFilesShown', value.toString())
  }
  const setAreFileExtensionsShown = (value: boolean) => {
    areFileExtensionsShown.value = value
    window.localStorage.setItem('oc_fileExtensionsShown', value.toString())
  }
  const setAreWebDavDetailsShown = (value: boolean) => {
    areWebDavDetailsShown.value = value
    window.localStorage.setItem('oc_webDavDetailsShown', value.toString())
  }

  const loadIndicators = (space: SpaceResource, id: string) => {
    const files = unref(resources).filter((f) => [f.id, f.parentFolderId].includes(id))
    for (const resource of files) {
      const indicators = getIndicators({
        space,
        resource,
        ancestorMetaData: unref(ancestorMetaData),
        user: userStore.user
      })
      if (!indicators.length && !resource.indicators.length) {
        continue
      }

      updateResourceField({ id: resource.id, field: 'indicators', value: indicators })
    }
  }

  const setAncestorMetaData = (value: AncestorMetaData) => {
    ancestorMetaData.value = value
  }

  const updateAncestorField = <
    T extends AncestorMetaDataValue,
    K extends keyof AncestorMetaDataValue
  >({
    path,
    field,
    value
  }: {
    path: T['path']
    field: K
    value: T[K]
  }) => {
    const resource = unref(ancestorMetaData)[path] ?? null
    if (resource) {
      resource[field] = value
    }
  }

  const loadAncestorMetaData = ({
    folder,
    space,
    client,
    signal
  }: {
    folder: Resource
    space: SpaceResource
    client: WebDAV
    signal?: AbortSignal
  }) => {
    const data: AncestorMetaData = {
      [folder.path]: {
        id: folder.fileId,
        shareTypes: folder.shareTypes,
        parentFolderId: folder.parentFolderId,
        spaceId: space.id,
        path: folder.path
      }
    }
    const promises = []
    const davProperties = [DavProperty.FileId, DavProperty.ShareTypes, DavProperty.FileParent]
    const parentPaths = getParentPaths(folder.path)
    const spaces = spacesStore.spaces

    const getMountPoints = () =>
      spaces.filter(
        (s) =>
          isMountPointSpaceResource(s) && extractStorageId(s.root.remoteItem.rootId) === space.id
      )

    let fullyAccessibleSpace = true
    if (configStore.options.routing.fullShareOwnerPaths) {
      // keep logic in sync with "isResourceAccessible" from useGetMatchingSpace
      const projectSpace = spaces.find((s) => isProjectSpaceResource(s) && s.id === space.id)
      fullyAccessibleSpace = space.isOwner(userStore.user) || projectSpace?.isMember(userStore.user)
    }

    for (const path of parentPaths) {
      const cachedData = unref(ancestorMetaData)[path] ?? null
      if (cachedData?.spaceId === space.id) {
        data[path] = cachedData
        continue
      }

      // keep logic in sync with "isResourceAccessible" from useGetMatchingSpace
      if (
        !fullyAccessibleSpace &&
        !getMountPoints().find((m) => path.startsWith(m.root.remoteItem.path))
      ) {
        // no access to the parent resource
        break
      }

      promises.push(
        client
          .listFiles(space, { path }, { depth: 0, davProperties, signal })
          .then(({ resource }) => {
            data[path] = {
              id: resource.fileId,
              shareTypes: resource.shareTypes,
              parentFolderId: resource.parentFolderId,
              spaceId: space.id,
              path
            }
          })
      )
    }

    return Promise.all(promises).then(() => {
      if (!Object.keys(data).includes('/')) {
        // add space as root element
        const cachedRoot = unref(ancestorMetaData)['/']
        if (cachedRoot?.spaceId === space.id) {
          data['/'] = cachedRoot
        } else {
          const { parentFolderId } = Object.values(data)[0]
          if (parentFolderId) {
            const space = spacesStore.spaces.find(({ id }) => parentFolderId.startsWith(id))
            if (space) {
              data['/'] = {
                id: space.id,
                shareTypes: space.shareTypes,
                parentFolderId: space.id,
                spaceId: space.id,
                path: '/'
              }
            }
          }
        }
      }

      setAncestorMetaData(data)
    })
  }

  const getAncestorById = (id: string) => {
    return Object.values(unref(ancestorMetaData)).find((a) => id === a.id)
  }

  return {
    resources,
    currentFolder,
    activeResources,
    totalResourcesCount,

    setResources,
    removeResources,
    clearResources,
    upsertResource,
    upsertResources,
    updateResourceField,

    setCurrentFolder,
    clearCurrentFolder,

    initResourceList,
    clearResourceList,

    selectedIds,
    latestSelectedId,
    selectedResources,
    setSelection,
    addSelection,
    removeSelection,
    toggleSelection,
    resetSelection,
    setLastSelectedId,

    areHiddenFilesShown,
    areFileExtensionsShown,
    areWebDavDetailsShown,
    setAreHiddenFilesShown,
    setAreFileExtensionsShown,
    setAreWebDavDetailsShown,

    loadIndicators,

    ancestorMetaData,
    setAncestorMetaData,
    updateAncestorField,
    loadAncestorMetaData,
    getAncestorById
  }
})

export type ResourcesStore = ReturnType<typeof useResourcesStore>
