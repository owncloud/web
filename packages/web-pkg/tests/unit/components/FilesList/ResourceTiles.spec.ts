import { defaultComponentMocks, defaultPlugins, mount } from 'web-test-helpers'
import ResourceTiles from '../../../../src/components/FilesList/ResourceTiles.vue'
import { sortFields } from '../../../../src/helpers/ui/resourceTiles'
import { Resource, ResourceIndicator, SpaceResource } from '@ownclouders/web-client'
import { mock } from 'vitest-mock-extended'
import { ComponentPublicInstance, computed } from 'vue'
import { extractDomSelector } from '@ownclouders/web-client'
import OcDrop from 'design-system/src/components/OcDrop/OcDrop.vue'
import { useCanBeOpenedWithSecureView } from '../../../../src/composables/resources'
import { displayPositionedDropdown } from '../../../../src/helpers/contextMenuDropdown'

vi.mock('../../../../src/helpers/contextMenuDropdown')
vi.mock('../../../../src/composables/viewMode', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useTileSize: vi.fn().mockReturnValue({
    calculateTileSizePixels: vi.fn().mockImplementation((viewSize: number) => 100 * viewSize)
  })
}))

const mockUseEmbedMode = vi.fn().mockReturnValue({ isEnabled: computed(() => false) })
vi.mock('../../../../src/composables/embedMode', () => ({
  useEmbedMode: vi.fn().mockImplementation(() => mockUseEmbedMode())
}))

vi.mock('../../../../src/composables/resources', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useCanBeOpenedWithSecureView: vi.fn()
}))

vi.mock('../../../../src/composables/actions/files', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useFileActions: vi.fn().mockReturnValue({
    getDefaultAction: vi.fn().mockReturnValue({ handler: vi.fn() })
  })
}))

const spacesResources = [
  {
    id: '1',
    name: 'Space 1',
    path: '',
    type: 'space',
    isFolder: true,
    indicators: [],
    getDriveAliasAndItem: () => '1'
  },
  {
    id: '2',
    name: 'Space 2',
    path: '',
    type: 'space',
    isFolder: true,
    indicators: [],
    getDriveAliasAndItem: () => '2'
  }
] as unknown as SpaceResource[]

const resources = [
  {
    id: 'forest',
    driveId: 'forest',
    name: 'forest.jpg',
    path: 'images/nature/forest.jpg',
    extension: 'jpg',
    thumbnail: 'https://cdn.pixabay.com/photo/2015/09/09/16/05/forest-931706_960_720.jpg',
    isFolder: false,
    indicators: [] as ResourceIndicator[],
    type: 'file',
    tags: ['space', 'tag', 'moon'],
    size: '111000234',
    hidden: false,
    syncEnabled: true,
    outgoing: false,
    canRename: vi.fn(),
    getDomSelector: () => extractDomSelector('forest'),
    canDownload: () => true
  } as Resource
]

describe('ResourceTiles component', () => {
  const originalGetElementById = document.getElementById
  const originalGetComputedStyle = window.getComputedStyle
  beforeEach(() => {
    const mockElement = {
      clientWidth: 800
    } as HTMLElement
    document.getElementById = vi.fn((id) => {
      if (id === 'tiles-view') {
        return mockElement
      }
      return originalGetElementById.call(document, id)
    })
    window.getComputedStyle = vi.fn().mockImplementation(() => {
      return {
        getPropertyValue: (propName: string) => {
          switch (propName) {
            case '--oc-size-tiles-default':
              return '9rem'
            case '--oc-size-tiles-resize-step':
              return '9rem'
            default:
              return originalGetComputedStyle(document.documentElement).getPropertyValue(propName)
          }
        },
        fontSize: '14px'
      }
    })
  })
  it('renders an array of spaces correctly', async () => {
    const { wrapper } = getWrapper({ props: { resources: spacesResources } })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders a footer slot', () => {
    const { wrapper } = getWrapper({ slots: { footer: 'Hello, ResourceTiles footer!' } })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('file click', () => {
    it('emits fileClick event upon click on tile', async () => {
      const { wrapper } = getWrapper({ props: { resources } })
      await wrapper.find('.oc-tiles-item .oc-resource-name').trigger('click')
      expect(
        wrapper.emitted<{ resources: Resource[] }[]>('fileClick')[0][0].resources[0].name
      ).toMatch('forest.jpg')
    })

    it('does not emit fileClick event upon click on tile when embed mode is enabled', async () => {
      mockUseEmbedMode.mockReturnValue({
        isEnabled: computed(() => true)
      })
      const { wrapper } = getWrapper({ props: { resources } })
      await wrapper.find('.oc-tiles-item .oc-resource-name').trigger('click')
      expect(wrapper.emitted().fileClick).toBeUndefined()
    })

    it('does not emit fileClick event if file can not be opened via secure view', async () => {
      const { wrapper } = getWrapper({
        canBeOpenedWithSecureView: false,
        props: {
          resources: [{ ...resources[0], isFolder: false, canDownload: () => false }]
        }
      })
      await wrapper.find('.oc-tiles-item .oc-resource-name').trigger('click')
      expect(wrapper.emitted().fileClick).toBeUndefined()
    })
  })

  it('emits update:selectedIds event on resource selection and sets the selection', () => {
    const { wrapper } = getWrapper({
      props: {
        resources: spacesResources,
        selectedIds: [spacesResources[0].id]
      }
    })
    wrapper.vm.toggleSelection(spacesResources[0])
    expect(
      wrapper.findComponent({ name: 'resource-tile' }).props('isResourceSelected')
    ).toBeTruthy()
    expect(wrapper.emitted('update:selectedIds')).toBeTruthy()
  })

  describe('sorting', () => {
    it('renders the label of the first sort field as default', () => {
      const { wrapper } = getWrapper({ props: { sortFields } })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortFields[0].label)
    })
    it('renders the label of the current sort field as default', () => {
      const sortField = sortFields[2]
      const { wrapper } = getWrapper({
        props: {
          sortFields,
          sortBy: sortField.name,
          sortDir: sortField.sortDir
        }
      })
      expect(wrapper.find('#oc-tiles-sort-btn').text()).toEqual(sortField.label)
    })
    it('emits the "sort"-event', async () => {
      const index = 2
      const { wrapper } = getWrapper({ props: { sortFields } })
      ;(wrapper.vm.$refs.sortDrop as ComponentPublicInstance<typeof OcDrop>).tippy = {
        hide: vi.fn()
      }
      await wrapper.findAll('.oc-tiles-sort-list-item').at(index).trigger('click')
      expect(wrapper.emitted('sort')).toBeTruthy()
      expect(wrapper.emitted('sort')[0][0]).toEqual({
        sortBy: sortFields[index].name,
        sortDir: sortFields[index].sortDir
      })
    })
    describe('drag and drop', () => {
      it('emits the "update:selectedIds"-event on drag start', async () => {
        const { wrapper } = getWrapper()
        wrapper.vm.dragItem = mock<Resource>()
        await wrapper.vm.$nextTick()
        ;(wrapper.vm.$refs.ghostElementRef as ComponentPublicInstance<unknown>).$el = { style: {} }
        wrapper.vm.dragStart(mock<Resource>(), {
          dataTransfer: { setDragImage: vi.fn() }
        } as unknown as DragEvent)
        expect(wrapper.emitted('update:selectedIds')).toBeDefined()
      })
      it('emits the "fileDropped"-event on resource drop', () => {
        const { wrapper } = getWrapper()
        wrapper.vm.fileDropped(mock<Resource>(), { dataTransfer: {} } as DragEvent)
        expect(wrapper.emitted('fileDropped')).toBeDefined()
      })
    })
    describe('context menu', () => {
      it('triggers the positioned dropdown on click', async () => {
        const spyDisplayPositionedDropdown = vi.mocked(displayPositionedDropdown)
        const { wrapper } = getWrapper({ props: { resources } })
        const btn = wrapper.find('.resource-tiles-btn-action-dropdown')
        await btn.trigger('click')
        expect(spyDisplayPositionedDropdown).toHaveBeenCalled()
      })
      it('does not show for disabled resources', () => {
        const { wrapper } = getWrapper({
          props: { resources: [{ ...resources[0], processing: true }] }
        })
        expect(wrapper.find('.resource-tiles-btn-action-dropdown').exists()).toBeFalsy()
      })
    })
    describe('checkboxes', () => {
      it('update the selected ids on click', async () => {
        const { wrapper } = getWrapper({ props: { resources } })
        const checkbox = wrapper.find('input[type="checkbox"]')
        await checkbox.trigger('click')
        expect(wrapper.emitted('update:selectedIds')).toBeTruthy()
      })
      it('are disabled for disabled resources', () => {
        const { wrapper } = getWrapper({
          props: { resources: [{ ...resources[0], processing: true }] }
        })
        const checkbox = wrapper.find('input[type="checkbox"]')
        expect(Object.keys(checkbox.attributes())).toContain('disabled')
      })
    })
  })

  it.each([
    { viewSize: 1, expected: 'xlarge' },
    { viewSize: 2, expected: 'xlarge' },
    { viewSize: 3, expected: 'xxlarge' },
    { viewSize: 4, expected: 'xxlarge' },
    { viewSize: 5, expected: 'xxxlarge' },
    { viewSize: 6, expected: 'xxxlarge' }
  ])('passes the "viewSize" to the OcTile component', async (data) => {
    const { wrapper } = getWrapper({
      props: { resources: spacesResources, viewSize: data.viewSize }
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.findComponent({ name: 'resource-tile' }).props('resourceIconSize')).toEqual(
      data.expected
    )
  })

  function getWrapper({ props = {}, slots = {}, canBeOpenedWithSecureView = true } = {}) {
    const mocks = defaultComponentMocks()

    vi.mocked(useCanBeOpenedWithSecureView).mockReturnValue({
      canBeOpenedWithSecureView: () => canBeOpenedWithSecureView
    })

    return {
      wrapper: mount(ResourceTiles, {
        props: {
          lazy: false,
          viewSize: 1,
          ...props
        },
        slots: {
          ...slots
        },
        global: {
          plugins: [...defaultPlugins()],
          mocks: mocks,
          provide: mocks
        }
      })
    }
  }
})
