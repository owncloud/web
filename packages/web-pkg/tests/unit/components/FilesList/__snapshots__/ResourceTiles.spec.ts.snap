// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceTiles component > renders a footer slot 1`] = `
"<div data-v-67d39700="" id="tiles-view" class="oc-px-m oc-pt-l">
  <div data-v-67d39700="" class="oc-flex oc-flex-middle oc-mb-m oc-pb-s oc-tiles-controls"><span data-v-67d39700=""><input id="resource-table-select-all" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" disabled="" aria-label="Select all"> <!--v-if--></span>
    <!--v-if-->
  </div>
  <ul data-v-67d39700="" class="oc-list oc-my-rm oc-mx-rm oc-tiles oc-flex"> </ul>
  <!--v-if-->
  <div data-v-67d39700="" class="oc-tiles-footer">Hello, ResourceTiles footer!</div>
</div>"
`;

exports[`ResourceTiles component > renders an array of spaces correctly 1`] = `
"<div data-v-67d39700="" id="tiles-view" class="oc-px-m oc-pt-l">
  <div data-v-67d39700="" class="oc-flex oc-flex-middle oc-mb-m oc-pb-s oc-tiles-controls"><span data-v-67d39700=""><input id="resource-table-select-all" type="checkbox" name="checkbox" class="oc-checkbox oc-rounded oc-checkbox-l" aria-label="Select all"> <!--v-if--></span>
    <!--v-if-->
  </div>
  <ul data-v-67d39700="" class="oc-list oc-my-rm oc-mx-rm oc-tiles oc-flex">
    <li data-v-67d39700="" class="oc-tiles-item has-item-context-menu">
      <div data-v-67d39700="" class="oc-tile-card oc-card oc-card-default oc-rounded" data-item-id="1" draggable="false"><a target="_self" draggable="false" class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm" tabindex="-1" attrs="[object Object]"></a>
        <div class="oc-card-body oc-p-s">
          <div class="oc-flex oc-flex-between oc-flex-middle">
            <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
              <div class="oc-resource oc-text-overflow">
                <!--v-if-->
                <div class="oc-resource-details oc-text-overflow"><a target="_self" draggable="false" class="oc-text-overflow" attrs="[object Object]"></a>
                  <div class="oc-resource-indicators">
                    <!--v-if-->
                  </div>
                </div>
              </div>
            </div>
            <div class="oc-flex oc-flex-middle">
              <!-- Slot for indicators !-->
              <!-- Slot for individual actions -->
              <!-- Slot for contextmenu --> <button data-v-67d39700="" type="button" aria-label="Show context menu" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw resource-tiles-btn-action-dropdown" id="context-menu-trigger-1">
                <!--v-if-->
                <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
                <div id="context-menu-drop-1" class="oc-drop oc-box-shadow-medium oc-rounded oc-overflow-hidden">
                  <div class="oc-card oc-card-body oc-background-secondary oc-p-s"></div>
                </div>
              </button>
            </div>
          </div>
          <!--v-if-->
        </div>
      </div>
    </li>
    <li data-v-67d39700="" class="oc-tiles-item has-item-context-menu">
      <div data-v-67d39700="" class="oc-tile-card oc-card oc-card-default oc-rounded" data-item-id="2" draggable="false"><a target="_self" draggable="false" class="oc-card-media-top oc-flex oc-flex-center oc-flex-middle oc-m-rm" tabindex="-1" attrs="[object Object]"></a>
        <div class="oc-card-body oc-p-s">
          <div class="oc-flex oc-flex-between oc-flex-middle">
            <div class="oc-flex oc-flex-middle oc-text-truncate resource-name-wrapper">
              <div class="oc-resource oc-text-overflow">
                <!--v-if-->
                <div class="oc-resource-details oc-text-overflow"><a target="_self" draggable="false" class="oc-text-overflow" attrs="[object Object]"></a>
                  <div class="oc-resource-indicators">
                    <!--v-if-->
                  </div>
                </div>
              </div>
            </div>
            <div class="oc-flex oc-flex-middle">
              <!-- Slot for indicators !-->
              <!-- Slot for individual actions -->
              <!-- Slot for contextmenu --> <button data-v-67d39700="" type="button" aria-label="Show context menu" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw resource-tiles-btn-action-dropdown" id="context-menu-trigger-2">
                <!--v-if-->
                <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span>
                <div id="context-menu-drop-2" class="oc-drop oc-box-shadow-medium oc-rounded oc-overflow-hidden">
                  <div class="oc-card oc-card-body oc-background-secondary oc-p-s"></div>
                </div>
              </button>
            </div>
          </div>
          <!--v-if-->
        </div>
      </div>
    </li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
    <li data-v-67d39700="" class="ghost-tile" aria-hidden="true"></li>
  </ul>
  <!--v-if-->
  <div data-v-67d39700="" class="oc-tiles-footer"></div>
</div>"
`;
